### PR TITLE
TELCODOCS-757-AZURE

### PR DESCRIPTION
**Pull request by StephenJamesSmith**
[TELCODOCS-757](https://issues.redhat.com/browse/TELCODOCS-757) NVAI-e 2.1 Azure enablement

Version(s):
4.12 and future versions

Issue:
Jira: https://issues.redhat.com/browse/TELCODOCS-757

Docs preview:
https://55577--docspreview.netlify.app/openshift-enterprise/latest/machine_management/creating_machinesets/creating-machineset-azure.html#nvidia-gpu-aws-adding-a-gpu-node_creating-machineset-azure

Dev: @fabiendupont

QE: @wabouhamad

Dev and QE reviews are complete.

Note to Reviewer: the module "modules/nvidia-gpu-aws-deploying-the-node-feature-discovery-operator.adoc" is reused and has already been reviewed.

### CODE REVIEW

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
++
+[source,terminal]
+----
+$ oc get machineset -n openshift-machine-api
+----
++
+.Example output
++
+[source,terminal]
+----
```
Comment (wabouhamad) [inline modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:300]: Replace Line 300 to 318:
`
NAME                                             DESIRED   CURRENT   READY   AVAILABLE   AGE
clustername-n6n4r-nc4ast4-gpu-worker-centralus1   1         1         1       1           122m
clustername-n6n4r-worker-centralus1               1         1         1       1           8h
clustername-n6n4r-worker-centralus2               1         1         1       1           8h
clustername-n6n4r-worker-centralus3               1         1         1       1           8h
`

**Review by wabouhamad — CHANGES_REQUESTED** [review]
Please replace the lines shown with comments.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+myclustername-worker-centralus2   1         1         1       1           6h9m
+myclustername-worker-centralus3   1         1         1       1           6h9m
+----
+
+. Make a copy of one of the existing compute `MachineSet` definitions and output the result to a YAML file by running the following command.
+This will be the basis for the GPU-enabled compute machine set definition.
++
+[source,terminal]
+----
+$ oc get machineset -n openshift-machine-api myclustername-worker-centralus1 -o yaml > machineset-azure-ocp.4.12.2-020623.yaml
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:59]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+----
+
+. Create the new file by running the following command:
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:62]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
++
+[source,terminal]
+----
+$ cat machineset-azure-ocp.4.12.2-020623.yaml
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:66]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+          acceleratedNetworking: true
+          apiVersion: machine.openshift.io/v1beta1
+          credentialsSecret:
+            name: azure-cloud-credentials
+            namespace: openshift-machine-api
+          diagnostics: {}
+          image:
+            offer: ""
+            publisher: ""
+            resourceID: /resourceGroups/myclustername-rg/providers/Microsoft.Compute/galleries/gallery_liqcui_azocp12_n6n4r/images/myclustername-gen2/versions/latest
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:117]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+              storageAccountType: Premium_LRS
+            osType: Linux
+          publicIP: false
+          publicLoadBalancer: myclustername
+          resourceGroup: myclustername-rg
+          spotVMOptions: {}
+          subnet: myclustername-worker-subnet
+          userDataSecret:
+            name: worker-user-data
+          vmSize:
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:139]: Added.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+          zone: "1"
+status:
+  availableReplicas: 1
+  fullyLabeledReplicas: 1
+  observedGeneration: 1
+  readyReplicas: 1
+  replicas: 1
+----
+
+. Make a copy of the `machineset-azure-ocp.4.12.2-020623.yaml` file by running the following command:
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:150]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 8]
```diff
@@ -0,0 +1,530 @@
...
++
+[source,terminal]
+----
+$ cp machineset-azure-ocp.4.12.2-020623.yaml machineset-azure-_GPU_ocp.4.12.2-020623.yaml
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:154]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 9]
```diff
@@ -0,0 +1,530 @@
...
+----
+
+. Create the new file by running the following command:
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:157]: Removed.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 10]
```diff
@@ -0,0 +1,530 @@
...
+  observedGeneration: 1
+  readyReplicas: 1
+  replicas: 1
+----
+
+. To verify your changes, perform a `diff` of the original compute definition and the new GPU-enabled node definition by running the following command:
++
+[source,terminal]
+----
+$ diff machineset-azure-ocp.4.12.2-020623.yaml machineset-azure-_GPU_ocp.4.12.2-020623.yaml
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:249]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 11]
```diff
@@ -0,0 +1,530 @@
...
+---
+>       machine.openshift.io/cluster-api-machineset: myclustername-nc4ast4-gpu-worker-centralus1
+30c30
+<         machine.openshift.io/cluster-api-machineset: myclustername-worker-centralus1
+---
+>         machine.openshift.io/cluster-api-machineset: myclustername-nc4ast4-gpu-worker-centralus1
+67c67
+<           vmSize: Standard_D4s_v3
+---
+>           vmSize:
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:270]: Added.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 12]
```diff
@@ -0,0 +1,530 @@
...
+----
+
+. Create the GPU-enabled compute machine set from the definition file by running the following command:
++
+[source,terminal]
+----
+oc create -f machineset-azure-_GPU_ocp.4.12.2-020623.yaml
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:277]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc (hunk 1 above)
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:300]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 13]
```diff
@@ -0,0 +1,530 @@
...
+. Update the following fields in `machineset-azure-_GPU_ocp.4.12.2-020623.yaml`:
+Update the following fields in `machineset-azure-_GPU_ocp.4.12.2-020623.yaml`:
++
+* Change .metadata.name to a name containing `gpu`.
+
+* Change `.spec.selector.matchLabels["machine.openshift.io/cluster-api-machineset"]` to match the new .metadata.name.
+
+* Change `.spec.template.metadata.labels["machine.openshift.io/cluster-api-machineset"]` to match the new `.metadata.name`.
+
+* Change .spec.template.spec.providerSpec.value.`vmSize` to `Standard_NC24s_v3`.
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:387]: Changed.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 14]
```diff
@@ -0,0 +1,530 @@
...
++
+.Example `machineset-azure-ocp.4.12.2-020623.yaml` file
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:389]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 15]
```diff
@@ -0,0 +1,530 @@
...
+          acceleratedNetworking: true
+          apiVersion: machine.openshift.io/v1beta1
+          credentialsSecret:
+            name: azure-cloud-credentials
+            namespace: openshift-machine-api
+          diagnostics: {}
+          image:
+            offer: ""
+            publisher: ""
+            resourceID: /resourceGroups/myclustername-rg/providers/Microsoft.Compute/galleries/gallery_liqcui_azocp12_n6n4r/images/myclustername-gen2/versions/latest
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:437]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 16]
```diff
@@ -0,0 +1,530 @@
...
+              storageAccountType: Premium_LRS
+            osType: Linux
+          publicIP: false
+          publicLoadBalancer: myclustername
+          resourceGroup: myclustername-rg
+          spotVMOptions: {}
+          subnet: myclustername-worker-subnet
+          userDataSecret:
+            name: worker-user-data
+          vmSize: Standard_NC24s_v3
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:459]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 17]
```diff
@@ -0,0 +1,530 @@
...
+  observedGeneration: 1
+  readyReplicas: 1
+  replicas: 1
+----
+
+. Create the GPU-enabled compute machine set from the definition file by running the following command:
++
+[source,terminal]
+----
+oc create -f machineset-azure-_GPU_ocp.4.12.2-020623.yaml
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:474]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 18]
```diff
@@ -0,0 +1,530 @@
...
+[source,terminal]
+----
+$ oc -n openshift-machine-api get machines | grep gpu
+----
++
+.Example output
++
+[source,terminal]
+----
+myclustername-nc4ast4-gpu-worker-centralus1   Running   a2-highgpu-1g   us-central1   us-central1-a   121m
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:524]: Replaced.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 19]
```diff
@@ -0,0 +1,530 @@
...
+.Example output
++
+[source,terminal]
+----
+NAME                                   DESIRED   CURRENT   READY   AVAILABLE   AGE
+myclustername-worker-centralus1        1         1         1       1           8h
+myclustername-worker-centralus2        1         1         1       1           8h
+myclustername-worker-centralus3        1         1         1       1           8h
+----
+
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:278]: Moved.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 20]
```diff
@@ -0,0 +1,530 @@
...
+          vnet: myclustername-vnet
+          zone: "1"
+status:
+  availableReplicas: 1
+  fullyLabeledReplicas: 1
+  observedGeneration: 1
+  readyReplicas: 1
+  replicas: 1
+----
+
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:156]: Moved.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 21]
```diff
@@ -0,0 +1,430 @@
...
+//
+//  * machine_management/creating-machinesets/creating-machineset-azure.adoc
+
+:_content-type: PROCEDURE
+[id="nvidia-gpu-aws-adding-a-gpu-node_{context}"]
+= Adding a GPU node to an existing {product-title} cluster
+
+You can copy and modify a default compute machine set configuration to create a GPU-enabled machine set and machines for the Azure cloud provider.
+
+The following table lists the validated instance types:
```
Comment (fabiendupont) [inline modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:11]: According to [NVIDIA GPU Operator | Platform Support](https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/platform-support.html), the following instance types are supported: NC-T4-v3, NC-v3, ND-A100-v4.

I'd say that the [ND-100-V4](https://learn.microsoft.com/en-us/azure/virtual-machines/nda100-v4-series) is missing, no?

Comment (wabouhamad) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:11]: @fabiendupont QE has not tested with ND-100-V4, if your team has tested that instance on Azure, let's add it to the table

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 22]
```diff
@@ -0,0 +1,435 @@
...
+|V100
+|4
+|x86
+
+|`Standard_NC4as_T4_v3`
+|T4
+|1
+|x86
+
+|`ND A100 v4`
```
Comment (StephenJamesSmith) [inline modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:27]: @fabiendupont @wabouhamad Please provide the "NVIDIA GPU accelerator" and "Architecture" for ND A100 v4.

Comment (wabouhamad) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:27]: @StephenJamesSmith "NVIDIA GPU accelerator" is "A100" and Architecture is "x86".

Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:27]: Added.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc (hunk 21 above)
Comment (fabiendupont) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:11]: I did test it myself :)

**Review by fabiendupont — APPROVED** [review]
/lgtm

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 23]
```diff
@@ -0,0 +1,435 @@
...
+|A100
+|8
+|x86
+|===
+
+[NOTE]
+====
+By default, Azure subscriptions don't have a quota for the Azure instance types with GPU. Customers have to request a quota increase for the Azure instance families listed above.
```
Comment (mburke5678) [inline modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:35]: s/don't do not

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc [hunk 24]
```diff
@@ -0,0 +1,435 @@
...
+myclustername-master-0                              Running   Standard_D8s_v3        centralus   2      6h40m
+myclustername-master-1                              Running   Standard_D8s_v3        centralus   1      6h40m
+myclustername-master-2                              Running   Standard_D8s_v3        centralus   3      6h40m
+myclustername-nc4ast4-gpu-worker-centralus1-w9bqn   Running      centralus   1      21m
+myclustername-worker-centralus1-rbh6b               Running   Standard_D4s_v3        centralus   1      6h38m
+myclustername-worker-centralus2-dbz7w               Running   Standard_D4s_v3        centralus   2      6h38m
+myclustername-worker-centralus3-p9b8c               Running   Standard_D4s_v3        centralus   3      6h38m
+----
+
+. View the existing nodes, machines, and machine sets by running the following command. Note that each node is an instance of a machine definition with a specific Azure region and OpenShift Container Platform role.
```
Comment (mburke5678) [inline modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:337]: s/OpenShift Container Platform/{product-title}

**Comment by mburke5678** [comment]
@StephenJamesSmith A coupe of nits. Otherwise LGTM

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc (hunk 23 above)
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:35]: changed.

File: modules/nvidia-gpu-azure-adding-a-gpu-node.adoc (hunk 24 above)
Comment (StephenJamesSmith) [reply modules/nvidia-gpu-azure-adding-a-gpu-node.adoc:337]: changed.

**Comment by mburke5678** [comment]
@StephenJamesSmith Does this PR need a merge? It has a do-not-merge label

**Comment by StephenJamesSmith** [comment]
[mburke5678](https://github.com/mburke5678) -Do-not-Mege label removed. Sorry.

**Pull request merged**